### ISSUE TITLE
ai/rsc components streamed through createStreamableUI are mounted multiple times when updating the ui node (update/done)

### ISSUE THREAD

**Issue by marcoripa96** [vercel/ai#1257]
### Description

I was reading through the https://chat.vercel.ai/ github example and in particular I found this [useStreamableText](https://github.com/vercel/ai-chatbot/blob/fa9f0947f0a7983cf5022cbbc1416910349dd5e4/lib/hooks/use-streamable-text.ts) hook. I report the implementation here for clarity:

```tsx
import { StreamableValue, readStreamableValue } from 'ai/rsc'
import { useEffect, useState } from 'react'

export const useStreamableText = (
  content: string | StreamableValue<string>
) => {
  const [rawContent, setRawContent] = useState(
    typeof content === 'string' ? content : ''
  )

  useEffect(() => {
    (async () => {
      if (typeof content === 'object') {
        let value = ''
        for await (const delta of readStreamableValue(content)) {
          console.log(delta)
          if (typeof delta === 'string') {
            setRawContent((value = value + delta))
          }
        }
      }
    })()
  }, [content])

  return rawContent
}
```

When I used that hook in my project like in the following example:

```tsx
export function AssistantMessage({
  content,
}: {
  content: string | StreamableValue<string>;
}) {
  const text = useStreamableText(content);

  return (
    <div
      style={{ wordBreak: "break-word" }}
    >
      {text}
    </div>
  );
}

```

I noticed the UI flashing during the last part of the stream. I checked directly on https://chat.vercel.ai/ and I noticed the same behaviour.

 I tried printing to console `rawContent` and I noticed that just before the last delta received, rawContent becomes an empty string `""`.
 
```tsx
import { StreamableValue, readStreamableValue } from 'ai/rsc'
import { useEffect, useState } from 'react'

export const useStreamableText = (
  content: string | StreamableValue<string>
) => {
  const [rawContent, setRawContent] = useState(
    typeof content === 'string' ? content : ''
  )
  
  // before the last delta I get "", it should be impossibile
  console.log({ rawContent });

  useEffect(() => {
    (async () => {
      if (typeof content === 'object') {
        let value = ''
        for await (const delta of readStreamableValue(content)) {
          console.log(delta)
          if (typeof delta === 'string') {
            setRawContent((value = value + delta))
          }
        }
      }
    })()
  }, [content])

  return rawContent
}
```
 
 I ended up figuring out that the actual `<AssistantMessage />` component is mounting mutiple times and it happens whenever the ui node is updated or when the `done` function is called and causing that flashing issue in my code. I made a reproduction [here](https://codesandbox.io/p/devbox/elated-mayer-hqdjyq?file=%2Fapp%2Faction.tsx%3A17%2C19). Also not sure why in the codesanbox preview isn't really noticeable, but if you try looking at the preview in a new tab like [here](https://hqdjyq-3000.csb.app/), you can see the issue more clearly.

This happens on:

- next: 14.2.0-canary.48
- ai: 3.0.16

**Comment by shuding** [comment]
Yes this is something related to `Suspense` and `render()`. Thanks for reporting!

**Comment by marcoripa96** [comment]
Hi! Thanks for the clarification! I just wanted to point out that I am not using the `render()`. This is also happening using the primitives createStreamableUI and createStreamableValue

**Comment by unstubbable** [comment]
In the posted example you can work around the mentioned suspense issue (i.e. the whole promise chain is replayed when the fallback is replaced with the children) by skipping the `ui.update()` step:

```diff
   (async () => {
     await sleep(50);
     const textNode = createStreamableValue("");
-    ui.update(<AssistantMessageWithStreamableValue content={textNode.value} />);
+    ui.done(<AssistantMessageWithStreamableValue content={textNode.value} />);
 
     for await (const token of generateTokens(text)) {
       textNode.update(token);
     }
 
     textNode.done();
-    ui.done();
   })();
```

**Comment by aayush-kapoor** [comment]
Marking this as closed for now - if you face any regression or have a similar problem after migrating to the latest AI SDK version, please feel free to open a new issue (with possible reproduction)

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This reports a reproducible failure in existing AI SDK streaming behavior: UI updates unexpectedly remount client components, reset state, and duplicate effects.

### Evidence
- Updating or completing a `createStreamableUI` node remounts the streamed client component multiple times.
- The remount resets streamed text to an empty string, causing visible UI flashing, and can repeat side effects such as adding products to a cart.
- A maintainer confirmed the behavior was related to Suspense, and another commenter provided a workaround that avoids `ui.update()`.
- The issue was later closed with a request to report any recurrence on the latest SDK, not because the original behavior was deemed expected.
